### DISCUSSION TITLE
CUE Community Call 2022-10-04

### DISCUSSION THREAD

**Discussion by myitcv** [cue-lang/cue#1958] [Community Calls]
We are excited to confirm Tuesday 4th October at 1530 UTC for our first CUE Community Call!

The session will feature:

* Introductions to full time contributors to CUE.
* Upcoming releases.
* Recent progress.
* Current focus items and priorities.
* Announcements.
* Q&A / office hours.

### Meeting details

https://meet.google.com/wkj-erha-ekr

This session will also be recorded for those unable to join us live.

### CUE Community shared calendar

https://calendar.google.com/calendar/u/0?cid=Y19lNzkxMWQ5OWQ4ZGIyMmU2ZTVjMzhkMTVkNjY2ZTVlNjdiNWE5ODNkZWU4N2JmNTU2NDY3NzI1OGIxYjJhMTFhQGdyb3VwLmNhbGVuZGFyLmdvb2dsZS5jb20

### Recording

https://www.youtube.com/watch?v=Y3NXfX_o3a8

### Slides

https://docs.google.com/presentation/d/13hvds28CwmsaTRIYkF1SSlx6MY24cxQu74kCvb1GGlE/edit?usp=sharing

### Umbrella issues discussed on the call

See https://github.com/cue-lang/cue/discussions/1958#discussioncomment-3953276.

But in the meantime, please reply with any questions/topics you would like to see covered.

Thanks!

**Comment by myitcv** [comment]
Updated the description above with confirmation that we will be recording the session for those unable to join us live.

**Comment by asdine** [comment]
I'd love to hear you talk about your plans regarding:
- documentation
- tooling
- support for other languages

**Reply by myitcv** [reply]
Thanks for asking these. 

Regarding:

> support for other languages

Please can you expand a bit on what you are trying to do here? There are a number of possible interpretations when it comes to CUE, so I want to make sure we answer your specific case (whilst touching more broadly on the others for completeness).

**Reply by nyarly** [reply]
I'd love to see CUE embedded in Rust e.g. for configuration or API management

**Reply by addreas** [reply]
I've been thinking a bit about language support as well. Would be interesting to hear thoughts on embedding via C/wasm compared to multiple independent implementations. Embedding is probably quicker and easier, but often becomes a bit clunky as well.

Also, some deeper docs/spec for the ADT as a reference for an implementation is something I've been missing, but that might belong further down the roadmap.

**Reply by myitcv** [reply]
Thanks both!

**Comment by nyarly** [comment]
You'll not be surprised to hear that I'm keenly interested to hear about performance :)

**Comment by jpluscplusm** [comment]
I'd be really interested to hear about anything and everything that folks are doing or discovering with CUE in the IaC ecosystem. 

Random ideas I'm just throwing out there:

- HCL support
- working with Terraform JSON in CUE
- any solutions on the horizon to the perf issues experienced when using @jlongtine's Cfn-cue (sp?)
- Any Azure ARM users out there specifying templates in CUE?
- The DevX tension between schema creation and /abstraction/ creation: where's the true value delivered? Who's doing what, and with what level of success?

Indeed: anything IaC or IaC-adjacent!

**Comment by cipriancraciun** [comment]
In my experiments with CUE I've hit two major (related) blocking issues and I would like to know if they are on the roadmap or something completely out-of-scope:

* scalability with regard to complexity -- what is the level of complexity that is considered "normal";  for example I've tried to define a quite complex "cron-like schedule" JSON object (supporting time ranges, time zones, etc.) that used quite a lot of fine-grained definitions, and when I added for example an explicit list of all time zones (~200 perhaps constant alternatives), the interpreter wouldn't terminate anymore;

* scalability with regard default values -- sometimes you want to force CUE to explicitly fill in all optional values with their defaults (so that you can use the output of CUE in another program without also implementing the defaults logic there);  however this has increased the evaluation time almost 10x;

Additionally I'm also interested in:
* support for other programming language integration (especially Rust or WASM);
* sandboxing of untrusted source files;  does CUE intend to provide support for sandboxing, or should one enforce sandboxing from the OS with things such as `seccomp` or even a lightweight VM (like Firecracker);

Last but not least, I would also like to touch upon the thorny subject of governance:  from what I gather CUE was originally developed by/at Google (?), however the old repository (<https://cue.googlesource.com/cue/> and <https://github.com/cuelang/cue>) seems to be different than the new one (<https://github.com/cue-lang/cue>).  So, especially given Google's track record of hyping a technology just to abandon it two years later, how reliant is CUE on Google's money with regard to development?

----

P.S. I understand that this is an open-source project and that I shouldn't expect anything from its developers, especially since I don't chip-in for its development.  But at the moment I'm trying to investigate possible solutions for an early startup, and thus at the moment I don't have the money to actively support the project.  But on the other hand relying on CUE will impact the future of my project, thus it would be helpful to know that other companies are able to support this project.  And perhaps in the future if I can I could contribute back, but I need to get there.  :)

(I know, the open-source chicken-and-egg-problem:  one needs money to get to a stable open-source project, but for that one needs users, which in turn requires a stable project to begin with.)  :) :)

**Reply by slewiskelly** [reply]
The reason why the project moved is detailed [here](https://github.com/cue-lang/cue/issues/1078):

> The GitHub organization https://github.com/cuelang is part of Google's Enterprise GitHub account. Moving the CUE project to https://github.com/cue-lang gives it a neutral home, and gives the project maintainers more flexibility to run the project and build for its long-term sustainability. We plan to move CUE to a foundation in the near future.

Not that it matters in the context of CUE, but I would try and avoid conflating Google's behavior with regards to retail / enterprise offerings and open-source. I would say Google's track record with continued investment in open-source projects is actually, on the whole, pretty good.

**Comment by myitcv** [comment]
Meeting details added in the description for this issue!

**Comment by tamalsaha** [comment]
Do you have any Google calendar where we can subscribe to?

**Reply by mvdan** [reply]
We've just created a public shared calendar with this first event :) Please try subscribing to it: https://calendar.google.com/calendar/u/0?cid=Y19lNzkxMWQ5OWQ4ZGIyMmU2ZTVjMzhkMTVkNjY2ZTVlNjdiNWE5ODNkZWU4N2JmNTU2NDY3NzI1OGIxYjJhMTFhQGdyb3VwLmNhbGVuZGFyLmdvb2dsZS5jb20

**Comment by eadlam** [comment]
I'm interested in updates on the Language Server.

**Comment by mvdan** [comment]
If anyone is having issues joining the Google Meet link above, please reply here or message me on either Twitter or Slack and I'll try to help.

**Comment by myitcv** [comment]
I've updated the description above with links to the recording on YouTube, as well as the slides. 

We will follow up with a link to all the umbrella issues discussed on the call.

**Comment by myitcv** [comment]
A bit later than planned, but here are links to umbrella issues discussed in Community Call 2022-05-04:

- Slide 7 - what breaking changes remain?
    - [language: placeholder for implementation of required fields proposal · Issue #2003 · cue-lang/cue](https://github.com/cue-lang/cue/issues/2003)
    - [language: make lists open by default · Issue #1999 · cue-lang/cue](https://github.com/cue-lang/cue/issues/1999)
    - [spec: need to specify iteration order of struct comprehension? · Issue #474 · cue-lang/cue](https://github.com/cue-lang/cue/issues/474)
    - [cue/load: do not include module ancestor files in the same package by default; make optional · Issue #2000 · cue-lang/cue](https://github.com/cue-lang/cue/issues/2000)
    - [Proposal: unify `float` and `int` as simply `number` · Issue #253 · cue-lang/cue](https://github.com/cue-lang/cue/issues/253)
    - [proposal: use ~ instead of =~ for regular expression constraints · Issue #2001 · cue-lang/cue](https://github.com/cue-lang/cue/issues/2001)
    - [cue/load: add CUE directive (or similar) to cue.mod · Issue #2004 · cue-lang/cue](https://github.com/cue-lang/cue/issues/2004)
- Slide 8 - release milestones
    - [Comprehension and cycle rework release milestone](https://github.com/cue-lang/cue/issues?q=is%3Aopen+is%3Aissue+milestone%3A%22v0.5.0+comprehension+rework%22)
    - Next releases will be created and announced via [CUE announcements - Discussion #2022](https://github.com/cue-lang/cue/discussions/2022)
- Slide 10 - Recent progress
    - [Comprehension and cycle rework release milestone](https://github.com/cue-lang/cue/issues?q=is%3Aopen+is%3Aissue+milestone%3A%22v0.5.0+comprehension+rework%22)
    - [Required fields proposal v2 · Discussion #1951 · cue-lang/cue](https://github.com/cue-lang/cue/discussions/1951)
    - [language: function signatures · Issue #2007 · cue-lang/cue](https://github.com/cue-lang/cue/issues/2007)
    - [Go release support policy · Discussion #1937 · cue-lang/cue](https://github.com/cue-lang/cue/discussions/1937)
- cuelang.org redesign
    - [Add support for private registrations · Issue #111 · cue-unity/unity](https://github.com/cue-unity/unity/issues/111)
- Slide 14 - Marcel focus for the rest of 2022
    - [Comprehension and cycle rework release milestone](https://github.com/cue-lang/cue/issues?q=is%3Aopen+is%3Aissue+milestone%3A%22v0.5.0+comprehension+rework%22)
    - [evaluator: disjunction performance improvements · Issue #2002 · cue-lang/cue](https://github.com/cue-lang/cue/issues/2002)
    - [evaluator: general performance enhancements · Issue #804 · cue-lang/cue](https://github.com/cue-lang/cue/issues/804)
    - [cuelang.org: new language guide · Issue #2012 · cue-lang/cue](https://github.com/cue-lang/cue/issues/2012)
    - [cue-unity/unity: unity - run experiments/regression tests on CUE modules](https://github.com/cue-unity/unity)
    - [evaluator: adding stats collection · Issue #2019 · cue-lang/cue](https://github.com/cue-lang/cue/issues/2019)
- Slide 16 - Paul focus for rest of 2022
    - [x/tools: Language Server Protocol Support · Issue #142 · cue-lang/cue](https://github.com/cue-lang/cue/issues/142)
    - [CUE announcements · Discussion #2022 · cue-lang/cue](https://github.com/cue-lang/cue/discussions/2022)
- Slide 18 - Carmen focus for rest of 2022
    - [cuelang.org: documentation and content updates · Issue #2008 · cue-lang/cue](https://github.com/cue-lang/cue/issues/2008)
- Slide 20 - Roger focus for rest of 2022
    - [CUE package management and modules · Issue #2032 · cue-lang/cue](https://github.com/cue-lang/cue/issues/2032)
    - [language: function signatures · Issue #2007 · cue-lang/cue](https://github.com/cue-lang/cue/issues/2007)
- Slide 22 - Daniel focus for rest of 2022
    - [Add support for private registrations · Issue #111 · cue-unity/unity](https://github.com/cue-unity/unity/issues/111)
    - [support running Go tests in projects (I909e547b) · Gerrit Code Review](https://review.gerrithub.io/c/cue-unity/unity/+/543960)
- Slide 24 - Aram focus for the rest of 2022
    - [CUE and WebAssembly (WASM) · Issue #2035 · cue-lang/cue](https://github.com/cue-lang/cue/issues/2035)
    - [language: function signatures · Issue #2007 · cue-lang/cue](https://github.com/cue-lang/cue/issues/2007)
    - [Denotational Semantics for CUE · Issue #2036 · cue-lang/cue](https://github.com/cue-lang/cue/issues/2036)
- Slide 28 - Roadmap 2023
    - [Proposal: CUE Querying extension · Issue #165 · cue-lang/cue](https://github.com/cue-lang/cue/issues/165)
    - [tooling: native embed and export · Issue #2031 · cue-lang/cue](https://github.com/cue-lang/cue/issues/2031)
    - [encoding: support for more encodings/adapters · Issue #2038 · cue-lang/cue](https://github.com/cue-lang/cue/issues/2038)
    - [cue: improve error messages · Issue #52 · cue-lang/cue](https://github.com/cue-lang/cue/issues/52)
    - [`cue test` support · Issue #209 · cue-lang/cue](https://github.com/cue-lang/cue/issues/209)
    - [cmd/cue: add diff support · Issue #8 · cue-lang/cue](https://github.com/cue-lang/cue/issues/8)
    - [proposal: cmd/cue: rename 'cue cmd' to 'cue run' or 'cuerun' · Issue #1325 · cue-lang/cue](https://github.com/cue-lang/cue/issues/1325)
    - [Integrations with other languages (beyond Go) · Issue #2039 · cue-lang/cue](https://github.com/cue-lang/cue/issues/2039)
